fdbrpc/include/fdbrpc/AsyncFileWriteChecker.actor.h

```diff
@@ -0,0 +1,381 @@
+			for (uint32_t page : pages) {
+				self->writing.erase(page);
+			}
+			return r;
+		});
+	}
+
+	Future<Void> truncate(int64_t size) override {
+		auto self = Reference<AsyncFileWriteChecker>::addRef(this);
+		return map(m_f->truncate(size), [self, size](Void r) {
+			int maxFullPage = size / checksumHistoryPageSize;
+			int oldSize = self->lru.size();
+			self->lru.truncate(maxFullPage);
+			checksumHistoryBudget.get() += oldSize - self->lru.size();
+			return r;
+		});
+	}
+
+	Future<Void> sync() override {
+		auto self = Reference<AsyncFileWriteChecker>::addRef(this);
+		return map(m_f->sync(), [self](Void r) {
+			self->syncedTime = AsyncFileWriteChecker::transformTime(now());
+			return r;
+		});
+	}
+
+	Future<Void> flush() override { return m_f->flush(); }
+	Future<int64_t> size() const override { return m_f->size(); }
+	std::string getFilename() const override { return m_f->getFilename(); }
+	void releaseZeroCopy(void* data, int length, int64_t offset) override {
+		return m_f->releaseZeroCopy(data, length, offset);
+	}
+	int64_t debugFD() const override { return m_f->debugFD(); }
+
+	struct WriteInfo {
+		WriteInfo() : checksum(0), timestamp(0) {}
+		uint32_t checksum;
+		uint32_t timestamp; // keep a precision of ms
+	};
+
+	struct node {
+		uint32_t page;
+		WriteInfo writeInfo;
+		node *next, *prev;
+		node(uint32_t _page, WriteInfo _writeInfo) {
+			page = _page;
+			writeInfo = _writeInfo;
+			page = _page;
+			next = NULL;
+			prev = NULL;
+		}
+	};
+
+	class LRU {
+	private:
+		node* start;
+		node* end;
+		std::unordered_map<uint32_t, node*> m;
+		std::string fileName;
+		int maxFullPage;
+
+		void insertHead(node* n) {
+			n->next = start->next;
+			start->next->prev = n;
+			n->prev = start;
+			start->next = n;
+		}
+
+		void removeFromList(node* n) {
+			n->prev->next = n->next;
+			n->next->prev = n->prev;
+		}
+
+	public:
+		LRU(std::string _fileName) {
+			fileName = _fileName;
+			maxFullPage = 0;
+			start = new node(0, WriteInfo());
+			end = new node(0, WriteInfo());
+			start->next = end;
+			end->prev = start;
+		}
+
+		void update(uint32_t page, WriteInfo writeInfo) {
+			if (m.find(page) != m.end()) {
+				node* n = m[page];
+				removeFromList(n);
+				insertHead(n);
+				n->writeInfo = writeInfo;
+				return;
+			}
+			node* n = new node(page, writeInfo);
+			insertHead(n);
+			m[page] = n;
+			if (page >= maxFullPage) {
+				maxFullPage = page + 1;
+			}
+		}
+
+		int maxPage() { return maxFullPage; }
+
+		void truncate(int newMaxFullPage) {
+			for (int i = newMaxFullPage; i < maxFullPage; i++) {
+				remove(i);
+			}
+			if (maxFullPage > newMaxFullPage) {
+				maxFullPage = newMaxFullPage;
+			}
+		}
+
+		int size() { return m.size(); }
+
+		bool exist(uint32_t page) { return m.find(page) != m.end(); }
+
+		WriteInfo find(uint32_t page) {
+			if (!exist(page)) {
+				TraceEvent(SevError, "CheckerTryFindingPageNotExist").log();
+				return WriteInfo();
+			}
+			return m[page]->writeInfo;
```

Comment:
This duplicates the work in `exist()`, i.e., `m.find(page)`. Better to keep the result of that:

```
auto it = m.find(page);
if (it != m.end()) { ... }
return it->second;
```